test/mongoose_subdomain_core_SUITE.erl

```diff
@@ -0,0 +1,567 @@
+    Handler = mongoose_packet_handler:new(?MODULE),
+    Subdomain1 = mongoose_subdomain_utils:get_fqdn(Pattern1, ?STATIC_DOMAIN),
+    Subdomain2 = mongoose_subdomain_utils:get_fqdn(Pattern1, hd(?DYNAMIC_DOMAINS)),
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?STATIC_HOST_TYPE,
+                                                                Pattern1, Handler)),
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE2,
+                                                                Pattern1, Handler)),
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                                Pattern2, Handler)),
+    ?assertEqual({ok, ?STATIC_HOST_TYPE},
+                 mongoose_subdomain_core:get_host_type(Subdomain1)),
+    ?assertEqual({ok, ?DYNAMIC_HOST_TYPE1},
+                 mongoose_subdomain_core:get_host_type(<<"some.fqdn">>)),
+    ?assertEqual({ok, ?DYNAMIC_HOST_TYPE2},
+                 mongoose_subdomain_core:get_host_type(Subdomain2)),
+    ?assertEqual({error, not_found},
+                 mongoose_subdomain_core:get_host_type(<<"unknown.subdomain">>)),
+    ?assertEqual({ok, #{host_type => ?STATIC_HOST_TYPE, subdomain_pattern => Pattern1,
+                        parent_domain => ?STATIC_DOMAIN, packet_handler => Handler,
+                        subdomain => Subdomain1}},
+                 mongoose_subdomain_core:get_subdomain_info(Subdomain1)),
+    ?assertEqual({ok, #{host_type => ?DYNAMIC_HOST_TYPE1, subdomain_pattern => Pattern2,
+                        parent_domain => undefined, packet_handler => Handler,
+                        subdomain => <<"some.fqdn">>}},
+                 mongoose_subdomain_core:get_subdomain_info(<<"some.fqdn">>)),
+    ?assertEqual({ok, #{host_type => ?DYNAMIC_HOST_TYPE2, subdomain_pattern => Pattern1,
+                        parent_domain => hd(?DYNAMIC_DOMAINS), packet_handler => Handler,
+                        subdomain => Subdomain2}},
+                 mongoose_subdomain_core:get_subdomain_info(Subdomain2)),
+    ?assertEqual({error, not_found},
+                 mongoose_subdomain_core:get_subdomain_info(<<"unknown.subdomain">>)),
+    ok.
+
+handles_domains_removal_during_subdomain_registration(_Config) ->
+    %% NumOfDomains is just some big non-round number to ensure that more than 2 ets
+    %% selections are done during the call to mongoose_domain_core:for_each_domain/2.
+    %% currently max selection size is 100 domains.
+    NumOfDomains = 1234,
+    NumOfDomainsToRemove = 1234 div 4,
+    NewDomains = [<<"dummy_domain_", (integer_to_binary(N))/binary, ".localhost">>
+                  || N <- lists:seq(1, NumOfDomains)],
+    [mongoose_domain_core:insert(Domain, ?DYNAMIC_HOST_TYPE1, dummy_src)
+     || Domain <- NewDomains],
+    meck:new(mongoose_domain_core, [passthrough]),
+    WrapperFn = make_wrapper_fn(NumOfDomainsToRemove * 2, NumOfDomainsToRemove),
```

Comment:
Injecting domain removal to `for_each_domain` seems very complicated and tightly coupled with the implementation... We can keep it as it is, but I am quite sure this will be difficult for anyone to understand.

Reply:
added extra comments